src/parser/sparqlParser/SparqlQleverVisitor.h

```diff
@@ -54,6 +54,11 @@ class SparqlQleverVisitor : public SparqlAutomaticVisitor {
   using PropertyList = ad_utility::sparql_types::PropertyList;
   using Any = antlrcpp::Any;
   size_t _blankNodeCounter = 0;
+  int64_t numInternalVariables_ = 0;
+  // A Stack of vector<Variable> that store the variables that are visible in a
+  // query body. Each element corresponds to nested Queries that the parse is
+  // currently parsing.
+  std::vector<std::vector<Variable>> visibleVariables_{{}};
```

Comment:
I will resolve this conversation. The implementation is now clean enough using this stack based approach, especially since there are only very few places in which we modify the stack.